src/Diagnostics.Scripts/CompilationService/Utilities/ScriptCompilation.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Diagnostics.Scripts.Utilities
+{
+    public static class ScriptCompilation
+    {
+        /// <summary>
+        /// Checks if compilation is required - if the script has changed or assembly is not loaded, compilation is done
+        /// </summary>
+        public static bool IsCompilationNeeded(string assembly, string script, string scriptETag, out Assembly loadedAssembly)
+        {
+            loadedAssembly = null;
+            return !IsSameScript(script, scriptETag) || !IsAssemblyLoaded(assembly, out loadedAssembly);
+        }
+
+        /// <summary>
+        /// Checks if the given assembly is already loaded
+        /// </summary>
+        public static bool IsAssemblyLoaded(string givenAssembly, out Assembly loadedAssembly)
+        {
+            loadedAssembly = null;
+            if(string.IsNullOrEmpty(givenAssembly))
+            {
+                return false;
+            }
+            Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies();
```

Comment:
Little bit concerned about this if we have to call this every time on compilation request from performance point of view.

Orthogonal to this, you can also create a in-process memory cache using dotnet core singleton service where you keep all the assemblies loaded in the cache (top 100 by time) and use that for optimization.

Reply:
Trying to understand this so correct me if I am wrong. Is this something similar to InvokeCache which gets Configured in Startup.cs? So I'd be creating a new service will cache the assemblies loaded and use that service to check if the assembly is present is the cache.

Reply:
Yes.. correct. A dotnet service runs in a singleton fashion and can be used by various controllers/components by simple dependency injection.
So You can create a service which hold latest 100 (or some number) of assemblies and check against that.